### PR TITLE
Bumped Lodash to v4

### PR DESCRIPTION
**Pull request by kunagpal**
Takes care of #89 

* Replaced JSHint and JSCS with ESLint and it's plugins
* Replaced deprecated method calls from Lodash v3 by their v4 equivalents
* Updated test config to work with ESLint
* Minor code style changes across the project

### CODE REVIEW

File: lib/collection/cookie.js [hunk 1]
```diff
@@ -267,11 +270,11 @@ _.extend(Cookie, /** @lends Cookie */ {
         key = split[0].trim();
         value = (split[1]) ? split[1].trim() : true;
 
-        if (_.isString(value) && '"' === value[0]) {
+        if (_.isString(value) && value[0] === QUOT) {
```
Comment (czardoz) [inline lib/collection/cookie.js]: Can we call it "QUOTE"?

File: lib/collection/cookie.js [hunk 2]
```diff
@@ -267,11 +270,11 @@ _.extend(Cookie, /** @lends Cookie */ {
...
             value = value.slice(1, -1);
         }
 
-        return { key: key, value: value };
+        return {key: key, value: value};
```
Comment (czardoz) [inline lib/collection/cookie.js]: @shamasis, is this the new ESLint configuration? I rather like the spaces there.

File: lib/collection/property-list.js [hunk 3]
```diff
@@ -155,34 +160,35 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
...
-        var match; // to be used if predicate is an ID
+        var match, // to be used if predicate is an ID
+            list = this;
 
-        !context && (context = this);
+        !context && (context = list);
         // if predicate is id, then create a function to remove that from array
         if (_.isString(predicate)) {
-            (match = this.one(predicate)) && (predicate = function (item) {
+            (match = list.one(predicate)) && (predicate = function (item) {
```
Comment (czardoz) [inline lib/collection/property-list.js]: Like we discussed the other day, this pattern can cause potential memory leaks. While it's not better or worse than a `.bind()`, there is no reason to change this here.

File: lib/collection/property-list.js [hunk 4]
```diff
@@ -155,34 +160,35 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
...
-                    delete this.reference[index];
+                if ((index = item[list._postman_propertyIndexKey]) && (index = String(index))) {
+                    list._postman_propertyIndexCaseInsensitive && (index = index.toLowerCase());
+                    delete list.reference[index];
                 }
                 delete item[__PARENT]; // unlink from its parent
                 return true;
             }
-        }, this);
+        });
```
Comment (czardoz) [inline lib/collection/property-list.js]: Use a `.bind()` at this place. That is what the old code used to do, and there's no reason to change it now.

File: lib/collection/property-list.js [hunk 5]
```diff
@@ -229,25 +235,29 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
      * Iterate on each item of this list
      */
     each: function (iterator, context) {
-        _.each(this.members, iterator, context || this.__parent);
+        _.forEach(this.members, iterator.bind(context || this));
```
Comment (czardoz) [inline lib/collection/property-list.js]: You're not binding the correct thing here. Look at the diff again.

File: lib/collection/property-list.js [hunk 6]
```diff
@@ -229,25 +235,29 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
...
     },
 
     /**
-     * @param {Function} rule
+     * @param {Function} iterator
      */
-    filter: function (rule, context) {
-        return _.filter(this.members, rule, context);
+    filter: function (iterator, context) {
+        return _.filter(this.members, iterator.bind(context || this));
```
Comment (czardoz) [inline lib/collection/property-list.js]: Do not introduce changes, the old code does not do a `||`.

File: lib/collection/property-list.js [hunk 7]
```diff
@@ -229,25 +235,29 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
...
      * Find an item within the item group
      *
-     * @param {Function} rule
+     * @param {Function} iterator
      * @param {Object} [context]
      * @returns {Item|ItemGroup}
      */
-    find: function (rule, context) {
-        return _.find(this.members, rule, context);
+    find: function (iterator, context) {
```
Comment (czardoz) [inline lib/collection/property-list.js]: Do not introduce changes, the old code does not do a `||` for context binding, and neither does it check whether the argument is a function.

File: lib/collection/property-list.js [hunk 8]
```diff
@@ -257,8 +267,7 @@ _.extend(PropertyList.prototype, /** @lends PropertyList.prototype */ {
      * @param context Optional context, defaults to the PropertyList itself.
      */
     map: function (iterator, context) {
-        !context && (context = this);
-        return _.map(this.members, iterator, context);
+        return _.map(this.members, iterator.bind(context || this));
```
Comment (czardoz) [inline lib/collection/property-list.js]: No `||`.

File: lib/collection/property.js [hunk 9]
```diff
@@ -52,8 +53,9 @@ _.inherit((
              * @note The property can also be present in the `postman_id` meta in case it is not specified in the
              * object. An auto-generated property is used wherever one is not specified
              */
-            id: src.id || (this._ ? this._['postman_id'] : undefined),
-            /*jshint +W069 */
+            id: src.id || (this._ ? this._postman_id : undefined),
```
Comment (czardoz) [inline lib/collection/property.js]: Why is this changed? /cc @shamasis

File: lib/collection/property.js [hunk 10]
```diff
@@ -172,6 +175,7 @@ _.extend(Property, /** @lends Property */ {
      * @returns {Object}
      */
     replaceSubstitutionsIn: function (obj, variables, mutate) {
+        var self = this;
```
Comment (czardoz) [inline lib/collection/property.js]: Use a `.bind()`, since that's the old behavior.

File: lib/collection/property.js [hunk 11]
```diff
@@ -180,9 +184,17 @@ _.extend(Property, /** @lends Property */ {
         // convert the variables to a substitutor object (will not reconvert if already substitutor)
         variables = Substitutor.box(variables, Substitutor.DEFAULT_VARS);
 
-        return _.merge(mutate ? obj : {}, obj, function (objectValue, sourceValue) {
-            return _.isString(sourceValue) ?  this.replaceSubstitutions(sourceValue, variables) : undefined;
-        }, this);
+        return _.mergeWith(mutate ? obj : {}, obj, function customizer(objectValue, sourceValue) {
+            if (!_.isString(sourceValue)) {
```
Comment (czardoz) [inline lib/collection/property.js]: Can you add a test case for this with a very deeply nested object containing a mixture of Arrays and other objects?

File: lib/collection/query-param.js [hunk 12]
```diff
@@ -136,7 +138,7 @@ _.extend(QueryParam, /** @lends QueryParam */ {
         if (!obj) { return ''; }
         var unparsed = obj.key || '';
         // if the value is not null, then we add the value
-        (obj.value != null) && (unparsed += ('=' + (obj.value || '')));
+        (obj.value !== null) && (obj.value !== undefined) && (unparsed += ('=' + (obj.value || '')));
```
Comment (czardoz) [inline lib/collection/query-param.js]: No changes in behavior please. Add a `todo` if you think this is incorrect, and we'll take it up in a future PR.

File: lib/collection/request-auth.js [hunk 13]
```diff
@@ -207,12 +215,12 @@ _.extend(RequestAuth, /** @lends RequestAuth */ {
             // todo: should we throw an error?
             return request;
         }
-        return handler.authorize.call(handler, request);
+        return handler.authorize(request);
```
Comment (czardoz) [inline lib/collection/request-auth.js]: Why is this changed?

File: lib/collection/request-auth/oauth1.js [hunk 14]
```diff
@@ -103,16 +105,14 @@ module.exports = {
                 system: true
             });
         }
+        else if (_.isString(request.method) && request.method.match(MATCH_PUTPOST) &&
```
Comment (czardoz) [inline lib/collection/request-auth/oauth1.js]: Not necessary to refactor code here. Revert this, and add a todo if you think this needs a change. We'll take it up in a future PR.

File: lib/collection/request-auth/oauth1.js [hunk 15]
```diff
@@ -103,16 +105,14 @@ module.exports = {
...
+                (request.body && request.body.mode === RequestBody.MODES.urlencoded)) {
+            _.forEach(signatureParams, function (param) {
+                request.body.urlencoded.add(param);
+            });
+        }
         else {
-            if (/PUT|POST/.test(request.method) &&
```
Comment (czardoz) [inline lib/collection/request-auth/oauth1.js]: Reinstate this code and ensure that the only changes in this file are space removals inside objects.

File: lib/collection/response.js [hunk 16]
```diff
@@ -17,22 +17,27 @@ var util = require('../util'),
...
     DOT = '.',
+
     /**
      * @private
      * @const
      * @type {string}
      */
     DEFAULT_RESPONSE_FILENAME = 'response',
 
+    MATCH_KNOWM_RESPONSE_ENCODING = /^(gzip|deflate)$/,
```
Comment (czardoz) [inline lib/collection/response.js]: Typo in the variable name.

File: lib/collection/response.js [hunk 17]
```diff
@@ -170,14 +177,15 @@ _.extend(Response.prototype, /** @lends Response.prototype */ {
      * @todo write unit tests
      */
     mime: function (contentType, contentDisposition) {
-        var detected = fileType(this.stream || this.body), // detect the mime from response body
+        var self = this,
```
Comment (czardoz) [inline lib/collection/response.js]: Unnecessary

File: lib/collection/response.js [hunk 18]
```diff
@@ -170,14 +177,15 @@ _.extend(Response.prototype, /** @lends Response.prototype */ {
...
+            detected = fileType(self.stream || self.body), // detect the mime from response body
             source = 'forced',
             mime;
 
         // if no overrides provided, we take the values from headers
-        !contentDisposition && (contentDisposition = this.headers.one('content-disposition'));
+        !contentDisposition && (contentDisposition = self.headers.one('content-disposition'));
```
Comment (czardoz) [inline lib/collection/response.js]: No need for `self` here

File: lib/collection/response.js [hunk 19]
```diff
@@ -215,8 +223,8 @@ _.extend(Response.prototype, /** @lends Response.prototype */ {
         }
 
         // we create the body string first from stream and then fallback to body
-        return 'data:' + mime._sanitisedContentType + ';base64, ' + (((this.stream != null) &&
-            util.bufferOrArrayBufferToBase64(this.stream)) || ((this.body != null) && util.btoa(this.body)) || E);
```
Comment (czardoz) [inline lib/collection/response.js]: Is this necessary?

File: lib/collection/response.js [hunk 20]
```diff
@@ -230,7 +238,8 @@ _.extend(Response.prototype, /** @lends Response.prototype */ {
             contentLength = Header.headerValue(this.headers, 'Content-Length');
 
         // if 'Content-Length' header is present and encoding is of type gzip/deflate
-        if (contentLength && /^(gzip|deflate)$/.test(contentEncoding) && util.isNumeric(contentLength)) {
+        if (contentLength && _.isString(contentEncoding) && contentEncoding.match(MATCH_KNOWM_RESPONSE_ENCODING) &&
```
Comment (czardoz) [inline lib/collection/response.js]: Typo in the variable name

File: lib/collection/url.js [hunk 21]
```diff
@@ -144,36 +152,36 @@ _.extend(Url.prototype, /** @lends Url.prototype */ {
...
         }
 
         if (this.host) {
-            rawUrl = rawUrl + this.getHost();
+            rawUrl += this.getHost();
         }
 
         if (this.port) {
-            rawUrl = rawUrl + ':' + this.port.toString();
+            rawUrl += (':' + this.port.toString(options));
```
Comment (czardoz) [inline lib/collection/url.js]: Where did the options come from?

File: lib/collection/url.js [hunk 22]
```diff
@@ -208,8 +216,10 @@ _.extend(Url.prototype, /** @lends Url.prototype */ {
      * @returns {String}
      */
     getQueryString: function (options) {
-        if (this.query.count()) {
-            return QueryParam.unparse(this.query.all(), options);
+        var self = this;
```
Comment (czardoz) [inline lib/collection/url.js]: Pointless

File: lib/collection/url.js [hunk 23]
```diff
@@ -144,36 +152,36 @@ _.extend(Url.prototype, /** @lends Url.prototype */ {
      *
      * @returns {string}
      */
-    toString: function () {
+    toString: function (options) {
```
Comment (czardoz) [inline lib/collection/url.js]: Revert all changes in this function, they're not needed.

File: lib/collection/request-auth.js (hunk 13 above)
Comment (shamasis) [inline lib/collection/request-auth.js]: This was a redundant .call. Eslint recommended this change afaik.

File: lib/collection/query-param.js (hunk 12 above)
Comment (shamasis) [inline lib/collection/query-param.js]: This is same behaviour na? Except that eslint did not suggest to use !=

File: lib/collection/cookie.js (hunk 2 above)
Comment (shamasis) [inline lib/collection/cookie.js]: Yes. I felt icky too and then was wondering that this was present in other places. I agree that we should go to one with spaces.

File: lib/collection/property-list.js (hunk 5 above)
Comment (shamasis) [inline lib/collection/property-list.js]: This is okay. I am not sure the old code was correct. We should keep the old code and add add a todo saying to check whether it's a bug in old code.

File: lib/collection/property-list.js (hunk 6 above)
Comment (shamasis) [inline lib/collection/property-list.js]: This change should ideally not cause any issue. What I would rather love is that code is modified to not do binding at all in case context is falsey. That should make things performant.

File: lib/collection/property.js (hunk 9 above)
Comment (shamasis) [inline lib/collection/property.js]: Good find. Seems that this is a bug introduced. It appears that the reference branch commit has been directly merged instead of being used as a reference.

File: lib/collection/request-auth/oauth1.js (hunk 15 above)
Comment (shamasis) [inline lib/collection/request-auth/oauth1.js]: This was marked as dead block by eslint. And it seems rightly so.

File: lib/collection/response.js (hunk 16 above)
Comment (shamasis) [inline lib/collection/response.js]: 👍🏼

File: lib/collection/response.js (hunk 19 above)
Comment (shamasis) [inline lib/collection/response.js]: Eslint recommendation to not use !=. I always felt that != null should be allowed. 😕
